server/app/mutations/stacks/sort_helper.rb

```diff
@@ -1,29 +1,79 @@
 module Stacks
   module SortHelper
 
+    class LinkError < StandardError
+      attr_accessor :service
+
+      def initialize(service, *args)
+        super(*args)
+        @service = service
+      end
+    end
+
+    # Sort services with a stack, such that services come after any services that they link to.
+    # This can only be used to sort services within the same stack. Links to services in other stacks are ignored.
+    #
     # @param [Array<GridService,Hash>]
+    # @raise [LinkError] service ... has missing links: ...
+    # @raise [LinkError] service ... has recursive links: ...
     # @return [Array<GridService,Hash>]
     def sort_services(services)
+      # Map of service name to array of deep links, including links of linked services
+      service_links = {}
+
+      # Build hash of service name to array of linked service names
+      # {service => [linked_service]}
+      services.each do |service|
+        service_links[service[:name]] = __links_for_service(service)
+      end
+
+      # Modify each service's array to add a direct reference to each linked service's own array of linked services
+      # {service => [linked_service, [linked_service_links]]}
+      service_links.each do |service, links|
+        links.dup.each do |linked_service|
+          if linked_service_links = service_links[linked_service]
+            service_links[service] << service_links[linked_service]
+          else
+            raise LinkError.new(service), "service #{service} has missing links: #{linked_service}"
+          end
+        end
+      end
+
+      # Flatten the nested references to arrays
+      # In case of recursive references, the Array#flatten! will fail with ArgumentError: tried to flatten recursive array
+      # {service => [linked_service, linked_service_link, ...]}
+      service_links.each do |service, links|
+        begin
+          service_links[service].flatten!
+        rescue ArgumentError
+          raise LinkError.new(service), "service #{service} has recursive links: #{links}"
```

Comment:
This is pretty random in terms of which service it gets raised for exactly, and what the links look like. It's not the service with the minimal link cycle, it could be any leaf service that links into a cycle of linked services.